File: solutions/ruby/knapsack/1/knapsack.rb
class Knapsack
    def initialize(max_weight)
        @max_weight = max_weight
    end
    def max_value(items)
        items = items.select { |item| item.weight <= @max_weight }
        return 0 if items.empty?
        items = items.sort_by { |item| item.value }
        items = items.reverse
        items.each_with_index do |item, index|
            if item.weight <= @max_weight
                return item.value
            end
        end
    end
end

File: solutions/javascript/knapsack/1/knapsack.js
export const knapsack = (maximumWeight, items) => {
  const dp = Array(items.length + 1)
  .fill(null)
  .map(() => Array(maximumWeight + 1).fill(0))

  for (let i = 1; i <= items.length; i++) {
    for (let w = 1; w <= maximumWeight; w++) {
      const item = items[i - 1]
      if (item.weight > w) {
        dp[i][w] = dp[i - 1][w]
      } else {
        dp[i][w] = Math.max(
          dp[i - 1][w],
          dp[i - 1][w - item.weight] + item.value
        )
      }
    }
  }

  return dp[items.length][maximumWeight]
};
